felix/bpf-gpl/tc.c

```diff
@@ -557,622 +569,843 @@ static CALI_BPF_INLINE void calico_tc_process_ct_lookup(struct cali_tc_ctx *ctx)
-				CALI_DEBUG("Returning related ICMP from workload to tunnel\n");
-			} else if (CALI_F_TO_HEP) {
-				/* Special case for ICMP error being returned by the host with the
-				 * backing workload into the tunnel back to the original host. It is
-				 * ICMP related and there is a return tunnel path. We need to change
-				 * both the source and destination at once.
-				 *
-				 * XXX the packet was routed to the original client as if it was XXX
-				 * DSR and we might not be on the right iface!!! Should we XXX try
-				 * to reinject it to fix the routing?
-				 */
-				CALI_DEBUG("Returning related ICMP from host to tunnel\n");
+	switch (nat_res) {
+		case NAT_ALLOW:
+			goto allow;
+		case NAT_ENCAP_ALLOW:
+			if (dnat_return_should_encap() && state->ct_result.ifindex_fwd != CT_INVALID_IFINDEX) {
+				rc = CALI_RES_REDIR_IFINDEX;
 			}
 
-			state->ip_src = HOST_IP;
-			state->ip_dst = state->ct_result.tun_ip;
-			goto nat_encap;
-		}
-
-		state->dport = state->post_nat_dport;
-		state->ip_dst = state->post_nat_ip_dst;
+			goto encap_allow;
+		default:
+			goto deny;
+	}
 
-		goto allow;
+allow:
+	if (state->ct_result.flags & CALI_CT_FLAG_SVC_SELF) {
+		CALI_DEBUG("Loopback SNAT\n");
+		seen_mark |=  CALI_SKB_MARK_MASQ;
+		CALI_DEBUG("marking CALI_SKB_MARK_MASQ\n");
+		fib = false; /* Disable FIB because we want to drop to iptables */
+	}
 
-	case CALI_CT_ESTABLISHED_SNAT:
-		CALI_DEBUG("CT: SNAT from %x:%d\n",
-				bpf_ntohl(state->ct_result.nat_ip), state->ct_result.nat_port);
+	if (CALI_F_TO_HEP && !skb_seen(ctx->skb) && is_dnat) {
+		struct cali_rt *r = cali_rt_lookup(state->post_nat_ip_dst);
 
-		if (dnat_return_should_encap() && state->ct_result.tun_ip) {
-			if (CALI_F_DSR && !(state->ct_result.flags & CALI_CT_FLAG_NP_NO_DSR)) {
-				/* SNAT will be done after routing, when leaving HEP */
-				CALI_DEBUG("DSR enabled, skipping SNAT + encap\n");
-				goto allow;
+		if (r && cali_rt_flags_local_workload(r->flags)) {
+			state->ct_result.ifindex_fwd = r->if_index;
+			CALI_DEBUG("NP local WL %x:%d on HEP\n",
+					bpf_htonl(state->post_nat_ip_dst), state->post_nat_dport);
+			ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+			fib = true; /* Enforce FIB since we want to redirect */
+		} else if (!r || cali_rt_flags_remote_workload(r->flags)) {
+			/* If there is no route, treat it as a remote NP BE */
+			if (CALI_F_LO || CALI_F_MAIN) {
+				state->ct_result.ifindex_fwd = NATIN_IFACE  ;
+				CALI_DEBUG("NP remote WL %x:%d on LO or main HEP\n",
+						bpf_htonl(state->post_nat_ip_dst), state->post_nat_dport);
+				ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+			}
+			ctx->state->flags |= CALI_ST_CT_NP_REMOTE;
+			fib = true; /* Enforce FIB since we want to redirect */
+		}
+	}
+
+encap_allow:
+	{
+		struct fwd fwd = {
+			.res = rc,
+			.mark = seen_mark,
+		};
+		fwd_fib_set(&fwd, fib);
+		return fwd;
+	}
+
+deny:
+	{
+		struct fwd fwd = {
+			.res = TC_ACT_SHOT,
+			.reason = ctx->fwd.reason,
+		};
+		return fwd;
+	}
+}
```

Comment:
This essentially splits the original code in do_nat() and post_nat() as we want to reuse the same code in two programs.